parastell/radial_distance_utils.py

```diff
@@ -0,0 +1,170 @@
+
+
+def calc_z_radius(point):
+    """
+    Calculate the distance from the z axis.
+
+    Arguments:
+        point (iterable of [x,y,z] float): point to find distance for
+    Returns:
+        (float): distance to z axis.
+    """
+    return (point[0] ** 2 + point[1] ** 2) ** 0.5
+
+
+def get_start_index(filament):
+    """
+    Find the index at which the filament crosses the xy plane on the OB
+    side of the filament
+
+    Arguments:
+        filament (list of list of float): List of points defining the filament
+
+    Returns:
+        max_z_index (int): index at which the filament crosses the xy plane
+    """
+    max_z_index = None
+    max_z_radius = None
+    for index, point in enumerate(filament[0:-1]):
+        next_point = filament[index + 1]
+        if point[2] / next_point[2] < 0:
+            z_radius = calc_z_radius(point)
+            if max_z_radius is None:
+                max_z_index = index
+                max_z_radius = z_radius
+
+            elif max_z_radius < z_radius:
+                max_z_index = index
+                max_z_radius = z_radius
+    return max_z_index
+
+
+def sort_filaments_toroidally(filaments):
+    """
+    Reorder filaments in order of increasing toroidal angle
+
+    Arguments:
+        filaments (list of list of list of float): List of filaments, which are
+            lists of points defining each filament.
+    Returns:
+        filaments (list of list of list of float): filaments in order of
+            increasing toroidal angle.
+    """
+    com_list = []
+
+    for fil in filaments:
+        com = np.average(fil, axis=0)
+        com_list.append(com)
+
+    com_list = np.array(com_list)
+    phi_arr = np.arctan2(com_list[:, 1], com_list[:, 0])
+
+    filaments = np.array([x for _, x in sorted(zip(phi_arr, filaments))])
+
+    return filaments
+
+
+def reorder_filaments(filaments):
+    """
+    Reorder the filaments so they start near the outboard xy plane crossing,
+    and begin by increasing z value.
+
+    Arguments:
+        filaments (list of list of list of float): List of filaments, which are
+            lists of points defining each filament.
+    Returns:
+        filaments (list of list of list of float): Reordered list of filaments,
+            suitable for building the magnet surface.
+    """
+    for filament_index, filament in enumerate(filaments):
+        # start the filament at the outboard
+        max_z_index = get_start_index(filament)
+        reordered_filament = np.concatenate(
+            [filament[max_z_index:], filament[0:max_z_index]]
+        )
+
+        # make sure z is increasing initially
+        if filament[max_z_index, 2] > filament[max_z_index + 1, 2]:
+            reordered_filament = np.flip(reordered_filament, axis=0)
+
+        # remove duplicate point since the start point might be different
+        _, idx = np.unique(reordered_filament, return_index=True, axis=0)
+        reordered_filament = reordered_filament[np.sort(idx)]
```

Comment:
If the duplicate is always (???) the start point before you reorder it, is it easier to remove it before you reorder it?

Reply:
yes that is easier, thank you